File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.*;
import java.util.*;

class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();

		st = new StringTokenizer(br.readLine());
		int V = Integer.parseInt(st.nextToken());
		int E = Integer.parseInt(st.nextToken());

		int start = Integer.parseInt(br.readLine());

		List<int[]>[] graph = new ArrayList[V+1];
		for (int i = 0; i <= V; i++) {
			graph[i] = new ArrayList<>();
		}

		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int src = Integer.parseInt(st.nextToken());
			int dst = Integer.parseInt(st.nextToken());
			int weight = Integer.parseInt(st.nextToken());
			graph[src].add(new int[] {dst, weight});
		}

		Solution solution = new Solution(V, E, graph, start);
		int[] result = solution.solve();
		for (int i = 1; i <= V; i++) {
			sb.append(result[i] == Integer.MAX_VALUE ? "INF" : result[i]).append("\n");
		}

		bw.write(sb.toString());
		bw.flush();
		bw.close();
		br.close();
	}

	private static class Solution {
		private int V;
		private int E;
		private int[] nodes;
		private List<int[]>[] graph;
		private int start;

		Solution(int V, int E, List<int[]>[] graph, int start) {
			this.V = V;
			this.E = E;
			this.nodes = new int[V + 1];
			Arrays.fill(this.nodes, Integer.MAX_VALUE);
			this.graph = graph;
			this.start = start;
		}

		public int[] solve() {
			Queue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(s -> s[1]));

			pq.add(new int[] {start, 0});
			this.nodes[start] = 0;

			while (!pq.isEmpty()) {
				int[] current = pq.poll();
				int node = current[0];
				int cost = current[1];

				if (cost > nodes[node]) continue;

				for (int[] edge : graph[node]) {
					int nextNode = edge[0];
					int weight = edge[1];

					if (cost + weight < nodes[nextNode]) {
						nodes[nextNode] = cost + weight;
						pq.add(new int[] {nextNode, nodes[nextNode]});
					}
				}
			}

			return nodes;
		}
	}
}

File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
from heapq import heappush, heappop

def solution():
    input = sys.stdin.readline
    
    # 1. 입력 받기
    nv, ne = map(int, input().split())
    k = int(input())
    
    # 2. 그래프 구조 최적화: 딕셔너리 대신 인접 리스트(List of Lists) 사용이 훨씬 빠름
    graph = [[] for _ in range(nv + 1)]
    for _ in range(ne):
        u, v, w = map(int, input().split())
        graph[u].append((v, w))
        
    # 3. 안전한 INF 값 설정
    INF = int(1e9) 
    distance = [INF] * (nv + 1)
    distance[k] = 0
    
    q = []
    heappush(q, (0, k))
    
    # 4. 표준 다익스트라 실행
    while q:
        dist, now = heappop(q)
        
        # 이미 처리된 (더 짧은 경로가 있는) 노드는 무시
        if distance[now] < dist:
            continue
            
        for nxt_node, weight in graph[now]:
            cost = dist + weight
            
            # 더 짧은 경로를 발견한 경우에만 큐에 삽입
            if cost < distance[nxt_node]:
                distance[nxt_node] = cost
                heappush(q, (cost, nxt_node))
                
    # 5. 결과 출력
    for i in range(1, nv + 1):
        if distance[i] == INF:
            print("INF")
        else:
            print(distance[i])

if __name__ == '__main__':
    solution()